### PR TITLE
#8 execute post build scripts

### LINKED ISSUE #8: Allow to execute custom logic after building an instance
When the build process finishes, we need to execute custom scripts which will attach proper prototype database volume to db container. It's also important that at this point the build instance ID is available as environment variable.

### PR DESCRIPTION
**Pull request by Kryniol**
fixes #8 

unfortunately I don't know how to handle frontend :(

### CODE REVIEW

File: server/modules/instantiation/jobs.js [hunk 1]
```diff
@@ -88,6 +88,13 @@ module.exports = function (config, portProvider, interpolationHelper, buildInsta
         }
     }
 
+    class RunAfterBuildTaskJob extends ComponentInstanceJob {
```
Comment (wit3k) [inline server/modules/instantiation/jobs.js]: jobs.js seems like a candidate for a separate directory/namespace/package
please create one and move all classes from this file to separate files

File: server/modules/instantiation/jobs.js [hunk 2]
```diff
@@ -417,7 +424,7 @@ module.exports = function (config, portProvider, interpolationHelper, buildInsta
                 buildInstance.log('Running docker-compose.');
 
                 var dockerCompose = spawn(
-                    'docker-compose', ['--file', dockerComposeFileName, 'up', '--abort-on-container-exit', '--no-color'],
+                    'docker-compose', ['--file', dockerComposeFileName, 'up', '-d', '--no-color'],
```
Comment (wit3k) [inline server/modules/instantiation/jobs.js]: seems like something that we want to move to configuration

File: server/modules/instantiation/jobs.js [hunk 3]
```diff
@@ -439,10 +446,6 @@ module.exports = function (config, portProvider, interpolationHelper, buildInsta
                 });
 
                 dockerCompose.on('exit', (code) => {
-                    reject(code);
-                });
-
-                dockerCompose.on('close', (code) => {
                     if (0 !== code) {
                         buildInstance.log('Failed to run docker-compose.');
                         reject(code);
```
Comment (wit3k) [inline server/modules/instantiation/jobs.js]: dont see a place where you catch rejected promisses

File: server/modules/instantiation/jobs.js [hunk 4]
```diff
@@ -457,6 +460,45 @@ module.exports = function (config, portProvider, interpolationHelper, buildInsta
...
+                        FEAT__INSTANCE_ID: componentInstance.buildInstance.id,
+                    }
+                )
+
+                componentInstance.log(`Running ${scriptPath}.`);
+
+                exec(scriptFullPath, {
+                    cwd: componentInstance.fullPath,
+                    env: env,
+                }, (error, stdout) => {
```
Comment (wit3k) [inline server/modules/instantiation/jobs.js]: not sure but this method call can be probably transofmed to another promise with bluebird

File: server/modules/instantiation/jobs.js [hunk 5]
```diff
@@ -457,6 +460,45 @@ module.exports = function (config, portProvider, interpolationHelper, buildInsta
...
+                    if (error) {
+                        componentInstance.error(`Failed to execute ${scriptPath}`);
+                        reject(error);
+
+                        return;
+                    }
+
+                    componentInstance.log(`Succeeded to execute ${scriptPath}`);
+                    componentInstance.log(stdout);
```
Comment (wit3k) [inline server/modules/instantiation/jobs.js]: I'd change order of those two

File: server/modules/instantiation/jobs.js [hunk 6]
```diff
@@ -457,6 +460,45 @@ module.exports = function (config, portProvider, interpolationHelper, buildInsta
...
+
+                componentInstance.log(`Running ${scriptPath}.`);
+
+                exec(scriptFullPath, {
+                    cwd: componentInstance.fullPath,
+                    env: env,
+                }, (error, stdout) => {
+                    if (error) {
+                        componentInstance.error(`Failed to execute ${scriptPath}`);
+                        reject(error);
```
Comment (wit3k) [inline server/modules/instantiation/jobs.js]: I'd change ordering so failure or success message is'nt wiped out by a long error/success logs from exec

File: server/modules/instantiation/jobs.js [hunk 7]
```diff
@@ -772,6 +815,19 @@ module.exports = function (config, portProvider, interpolationHelper, buildInsta
...
+        _.each(
+            buildInstance.componentInstances,
+            (componentInstance) => {
+                _.each(
+                    componentInstance.config.afterBuildScripts,
+                    (scriptPath) => {
+                        var job = new RunAfterBuildTaskJob(componentInstance, scriptPath);
+                        dependantJobsExecutor.add(job, [runDockerComposeJob]);
+                    }
+                );
```
Comment (wit3k) [inline server/modules/instantiation/jobs.js]:
```
_.each(
  componentInstance.config.afterBuildScripts,
  scriptPath => dependantJobsExecutor.add(new RunAfterBuildTaskJob(componentInstance, scriptPath), [runDockerComposeJob])
)
```

**Comment by mariusz-bak-xsolve-pl** [comment]
@Kryniol can you give some examples of after build script you would like to run?

**Comment by Kryniol** [comment]
@mariusz-bak-xsolve-pl as in #8, initially we need to attach a volume to newly created DB container and also configure Apache vhost along with proxy so that the app would be available under a subdomain. Later on we'll probably need some logic which would notify external webservices like Travis, Github or Jira that the environment is up and running.

**Comment by malef** [comment]
@Kryniol The reason why I'm asking is that I'm a bit reluctant to allow to execute _any_ script, and as I understood this one will be provided in repository, so there's virtually no control over what it can do. So it's a bit like using `exec` in PHP. That is why I decided to provide specific before build tasks, instead of allowing to run some bash script then.

Therefore I'm wondering if we can have some dedicated jobs that will do what you need and where we can validate somehow. Adding a subdomain would be one job type, attaching a volume - another one etc. 

What's your take on this?

**Comment by Kryniol** [comment]
I'd love to have it made this way as well but this requires a lot of additional work. So let's treat executing any script as MVP (because we need that feature urgently) and agree to improve the solution over time. What do you think?

**Comment by malef** [comment]
@Kryniol I get it. But let's agree that this will not be supported in future versions and it will be replaced later with some fine-grained commands, OK? To accept this we still would need frontend part though. I can try to find some time in the evening today to do it.

Am I right that the structure that you are adding to Mongo now look like the one given below?
```json
"afterBuildScripts": [
    "script1.sh",
    "script2.sh"
]
```

File: server/modules/instantiation/jobs.js (hunk 1 above)
Comment (Kryniol) [inline server/modules/instantiation/jobs.js]: I agree but let's do it in separate PR

File: server/modules/instantiation/jobs.js (hunk 2 above)
Comment (Kryniol) [inline server/modules/instantiation/jobs.js]: what exactly?

File: server/modules/instantiation/jobs.js (hunk 3 above)
Comment (Kryniol) [inline server/modules/instantiation/jobs.js]: I think they're supposed to fail but let's ask @mariusz-bak-xsolve-pl :)

**Comment by Kryniol** [comment]
@malef You're right, that's the proper structure. I agree this feature can be removed when we're done with more sophisticated solution.

File: server/modules/instantiation/jobs.js (hunk 2 above)
Comment (malef) [inline server/modules/instantiation/jobs.js]: I think it's fine as it is. Previously running `docker-compose` was the last stage so I didn't need to detach this.

File: server/modules/instantiation/jobs.js (hunk 6 above)
Comment (Kryniol) [inline server/modules/instantiation/jobs.js]: here I just add simple messsage that the script has failed and reject, nothing will be wiped out as I don't use stderr (not to duplicate error messages)

File: server/modules/instantiation/jobs.js (hunk 3 above)
Comment (malef) [inline server/modules/instantiation/jobs.js]: For now handling errors is really non-existent. But that something we should address in a separate PR after it is established how the information can be relayed to SPA (e.g. via websockets).

File: server/modules/instantiation/jobs.js (hunk 4 above)
Comment (Kryniol) [inline server/modules/instantiation/jobs.js]: the feature is straightforward - execute list of scripts one by one, without any nesting so I think promises won't bring any value here

File: server/modules/instantiation/jobs.js (hunk 1 above)
Comment (Kryniol) [inline server/modules/instantiation/jobs.js]: actually this one has been already reported in https://github.com/xsolve-pl/xsolve-feat/issues/13

**Comment by malef** [comment]
@Kryniol Sorry it's not merged yet, I still need to find time to handle frontend changes.
